### ISSUE TITLE
CommandButton: when ajax=false, it doesn't work when used inside composite component with a form 

### ISSUE THREAD

**Issue by newk5** [primefaces/primefaces#8801]
### Describe the bug

When using a composite component which has a h:form inside and using cc:renderFacet to put a p:commandButton with ajax=false inside, that page won't render with the following error:
```
FacesException: CommandButton : "j_id_7:j_id_8" must be inside a form element
    at org.primefaces.component.commandbutton.CommandButtonRenderer.buildRequest (CommandButtonRenderer.java:153)
```
I think this issue might be related to #8317 

### Reproducer

[primefaces_test.7z](https://github.com/primefaces/primefaces/files/8735132/primefaces_test.zip)

### Expected behavior

Primefaces should be able to detect that the p:commandButton is inside a h:form already even though the h:form is inside the composite component

### PrimeFaces edition

Community

### PrimeFaces version

11.0.0

### Theme

ALL?

### JSF implementation

All

### JSF version

ALL?

### Browser(s)

ALL?

**Comment by melloware** [comment]
So its the same issue the  `getClosestForm` walks up the tree but doesn't drill into Composite Components??

**Comment by newk5** [comment]
Yea seems like it. To be clear, this is when you use a  f:facet like:

![image](https://user-images.githubusercontent.com/32560510/169569957-e4bb4aec-2e2b-466c-a9ea-9bf967377496.png)

Do you know of any workaround I could try?

**Comment by melloware** [comment]
No so what happens is that code walks up the tree and find my:component but never finds an h:form.  Your only workaround right now is to wrap your `my:component` with the `h:form` so its outside of it would be my guess.

**Comment by newk5** [comment]
Hmm I see, unfortunately that's not an option for us. The h:form itself is inside "my:component" in many of our composite components.

**Comment by melloware** [comment]
Unfortunately as far as I can see that is the only workaround for now that would actually work.

**Comment by melloware** [comment]
It looks like OmniFaces does the same thing we are doing ` getClosestParent(component, UIForm.class)`...

```java
/**
	 * Returns from the given component the closest parent of the given parent type, or <code>null</code> if none
	 * is found.
	 * @param <C> The generic component type.
	 * @param component The component to return the closest parent of the given parent type for.
	 * @param parentType The parent type.
	 * @return From the given component the closest parent of the given parent type, or <code>null</code> if none
	 * is found.
	 */
	public static <C extends UIComponent> C getClosestParent(UIComponent component, Class<C> parentType) {
		UIComponent parent = component.getParent();

		while (parent != null && !parentType.isInstance(parent)) {
			parent = parent.getParent();
		}

		return parentType.cast(parent);
	}
```

**Comment by melloware** [comment]
Also lots of old posts with the general consensus is not to embed forms in the CC.

https://stackoverflow.com/questions/25795164/how-to-update-a-composite-component-form-from-another-composite-component

https://stackoverflow.com/questions/12952404/problems-with-composite-components-that-include-form

https://stackoverflow.com/questions/28201827/should-i-use-hform-in-composite-component-or-not

**Comment by newk5** [comment]
I think the issue when using h:form inside composite components is that you can very easily nest forms without noticing, but in my case there, I'm not doing that. There are many valid use cases to have a form within a composite component, we have many composite components like that. We use it to create predefined page layouts which we then insert the content from outside with f:facet: Just 1 example:

```html
<my:crud value="#{bean.myCrudController}">
   <f:facet name="dialog">
        content we insert here from the outside
    </f:facet>

</my:crud>
```
This crud composite component will then create something like:

```html
<h:form>
   <p:commandButton value="Create" onclick="PF('dlg').show()" />
   <p:dataTable >
       //dynamically built columns, with lazy loading, delete buttons, edit buttons etc...
  </p:dataTable >
</h:form>

<h:form id="dlg">
      <p:dialog position="top" header="Create/Edit" 
                  modal="true"
                  widgetVar="dlg" width="700" responsive="true" >
          <cc:renderFacet name="dialog" />
      </p:dialog>
</h:form>

```
It usually has more stuff in there but you get the point. In that case it even creates 2 forms inside as you can see. And it saves us a ton of time. We haven't had any issues with it so far except for this issue and #8317

**Comment by melloware** [comment]
Yep not arguing with you just saying it seems like its not a widely recommended practice.  The only way we can fix it is to modify that `getClosestForm` and it will fix both Menu and CommandButton

**Comment by melloware** [comment]
Let me take a shot at this ...

**Comment by newk5** [comment]
Yea I think its not recommended mostly because it can lead to accidently nesting forms. Also I recall @tandraschko said on #8317 that the server lookup could be made optional to make it work in a similar manner to p:ajax.

**Comment by tandraschko** [comment]
try it but i dont think we can easily support this.
the easiest way is to make it optional, log a warning as like with p:ajax and try to find a form, if required, on client side.

**Comment by newk5** [comment]
I should also mention that even in other situations PF doesn't detect forms when they're inside composite components and prints that log. I had to disable  that specific log message with logback because I could not see my server logs because they were being filled up completely with that message. Here's an example:

```
Component 'selectLists:fieldDatatype' should be inside a form or should reference a form via its form attribute. We will try to find a fallback form on the client side. 
```
![image](https://user-images.githubusercontent.com/32560510/169592701-62b71ab6-95da-451d-bf1a-70da43f20236.png)

**Comment by tandraschko** [comment]
yep, sure
that doesnt work in any case in PF when we look for a form

IMO we can even think about adding a "wont fix" label here but thats something you dont like ;D but the fallback + log thing is ok for me

**Comment by newk5** [comment]
Yea I dont mind the log spam too much since I can get around it by suppressing it with logback, that can be a WONTFIX, I'm fine with that, I just mentioned that because I thought it might be related to the issue of PF not being able to find a parent form even though a parent form exists.

My main issue is not being able to use a commandButton with ajax false or a menuItem inside a composite component with a h:form, which in my opinion should be supported.

**Comment by melloware** [comment]
OK so @tandraschko there is no easy fix unless I am mistaken.  In @newk5 use case above when it walks the hiearchy looking for `UIForm` it finds this..

```
UINamingContainer
HtmlBody
UIViewRoot
```

The first thing is the `UINamingContainer` which is your composite component.  You can't just walk `getChildren()` of that UINamingContainer looking for forms you would have to do a `visitTree`.  and then inside there if you found more UINamingContainer you would have to keep calling `visitTree` to try and get to a Form.

Unless I am mistaken which I might be!

**Comment by tandraschko** [comment]
Yeah
Thats very hard, and If you find one, you are not sure If its really the parent one
